Feat: allow dropping of USD/USDZ files on page to load them

hybridherbst opened the pull request:
### Description of Change(s)

Allows to quickly preview USDZ files and test the capabilities of the web port by dropping files onto the page.

~~Note that this is a DRAFT: the one piece that's still missing is unloading the existing file. Right now, each dragged file will be added to the scene.~~

<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

hybridherbst committed:
feat: allow dropping of USD/USDZ files on page to load them

hybridherbst commented on usd_for_web_demos/test.html:
```diff
@@ -19,7 +19,7 @@
     // Load the usd file
     let loadUSDPromise = new Promise( (resolve, reject) => {
       let req = new XMLHttpRequest();
-      req.open("GET", "/" + filename, true);
+      req.open("GET", filename, true);
```
This change is also part of #4

hybridherbst replied:
Note that this requires changing the links in index.html to include this slash (which is part of this PR).

hybridherbst commented on usd_for_web_demos/test.html:
```diff
@@ -38,9 +38,16 @@
+      let parts = filename.split('.');
+      let extension = parts[parts.length - 1];
+      extension = extension.split('?')[0];
+      let fileName = (Math.random() + 1).toString(36).substring(7);
+      // fileName = "input"; // TODO need to unload the old file again, then we don't need a random filename here
+      let inputFile = fileName + "." + extension;
```
This is using a random filename because otherwise the file already exists on subsequent loads.
The "old" file needs to be removed here, but I'm not sure how.

hybridherbst replied:
@kaischroeder would be looking for your input here especially - unloading the previous file is the main part still missing to make this useful.

kaischroeder replied:
is this just about the data file? In this case you could try unlinking it - this is using the Emscripten Filesystem API:

``Usd.FS.unlink('/' + inputFile);``


https://emscripten.org/docs/api_reference/Filesystem-API.html#FS.unlink

The only question is when :-) Doing it after calling driver.Draw() should be safe I guess.

hybridherbst replied:
I tried to unlink actually, but then got all kinds of exceptions from what I assume to be residuals of registering the previous file with .Draw. I didn't find a way to "unregister" it again.
Without that, both the old and new file will be rendered...

hybridherbst replied:
Got it to work! Key was to not only unlink the file but also remove the created objects from the Three.js scene graph.

kaischroeder commented on usd_for_web_demos/test.html:
```diff
@@ -52,12 +59,13 @@
-      camera.position.z = params.get('cameraZ') || 7;;
+      camera.position.z = params.get('cameraZ') || 7;
       camera.position.y = params.get('cameraY') || 7;
+      camera.position.y = params.get('cameraX') || 0;
```
typo - this should be camera.position.x = ...

kaischroeder commented on usd_for_web_demos/test.html:
```diff
@@ -151,6 +162,63 @@
+    function testAndLoadFile(file) {
+      let ext = file.name.split('.').pop();
+      console.log(file.name + ", " + file.size + ", " + ext);
+      if(ext == 'usd' || ext == 'usdz') {
```
we can add usda and usdc here

hybridherbst committed:
fix typo, add usda and usdc to extension filter

hybridherbst commented:
@kaischroeder I looked into this a bit more and while I can hack around and unlink the file plus stop rendering, it's not ideal - I think there's a missing method in the RenderDelegateInterface to clean up, e.g. to remove the meshes it has added into the three scene from the scene again.

It also seems that there's no scene hierarchy preserved from the USDZ right now in the ThreeJsRenderDelegate, it just adds everything as root-level meshes. I think it would be better to (Step 1) parent everything under one named Group (UsdStage?), and (Step 2) preserving the USD hierarchy would be great for future use.

(There's actually a little // FIXME comment in there)
`window.scene.add(this._mesh); // FIXME`

hybridherbst committed:
remove existing objects from scene graph and unlink loaded file when dropping a new one

hybridherbst committed:
fix links to test files, fix viewer page margins

hybridherbst committed:
feat: automatically adjust camera framing after loading objects

hybridherbst commented:
I changed a couple of things and can now properly unload models by
- grouping the scene objects under a common THREE.Group on loading
- clearing the group
- unlinking the old file

I also added automatic camera framing when loading models, which makes navigation a lot easier. I think this is a proper PR now and not just a draft, works pretty well on my end - you can test on https://prefrontalcortex.github.io/USD/usd_for_web_demos/test.html

kaischroeder commented:
This has been merged as #6
